cep-9999.md

```diff
@@ -0,0 +1,80 @@
+# CEP ???? - Names in conda packages and channels
+
+<table>
+<tr><td> Title </td><td> CEP ???? - Names in conda packages and channels </td>
+<tr><td> Status </td><td> Draft </td></tr>
+<tr><td> Author(s) </td><td> Jaime Rodríguez-Guerra &lt;jaime.rogue@gmail.com&gt; <br /> Matthew R. Becker &lt;becker.mr@gmail.com&gt;</td></tr>
+<tr><td> Created </td><td> Mar 11, 2025</td></tr>
+<tr><td> Updated </td><td> Mar 11, 2025</td></tr>
+<tr><td> Discussion </td><td> N/A </td></tr>
+<tr><td> Implementation </td><td> N/A </td></tr>
+</table>
+
+## Abstract
+
+This CEP aims to standardize names and other strings used throughout the conda ecosystem.
+
+## Specification
+
+### Artifacts
+
+The conda ecosystem distinguishes between two types of packages:
+
+- Regular package names: represented by a concrete, downloadable, extractable conda artifact.
+- Virtual package names: not backed by any physical artifact. They only exist on the client side.
+
+### Package names
+
+A regular package name MUST only consist of lowercase ASCII letters, numbers, hyphens, periods and underscores. It MUST start with a letter, a number, or a single underscore. It MUST NOT start with two consecutive underscores.
+
+Virtual package names MUST only consist of lowercase ASCII letters, numbers, hyphens, periods and underscores. They MUST start with two underscores.
+
+This means that regular package names are limited to those which match the following case-insensitive regex: `^(([a-z0-9])|([a-z0-9_](?!_)))[._-]?([a-z0-9]+(\.|-|_|$))*$`.
+
+Virtual package names MUST follow this other regex: `^__([a-z0-9]+(\.|-|_|$))+$`.
+
+### Version strings
+
+Version strings MUST only consist of digits, periods, lowercase ASCII letters, underscores, hyphens, plus signs and exclamation marks. Hyphens and underscores MAY be used interchangeably, but they MUST NOT be mixed in the same version string.
```

Comment:
We need to look at this carefully. Right now a lot of the conda code base uses `.rsplit("-", maxsplit=2)` to extract package name, version and build strings. Further, the VersionOrder class in conda maps all hyphens to underscores if only hyphens are in the version string. I have also never seen conda build output a package name with a dash in the version string. I think what happens is that hyphens are normalized to underscores if possible, but that hyphens are not allowed in version strings used in conda package names, the repodata, and basically any other place.

Reply:
I also noticed that trailing underscores are allowed but they are not used to generate empty version parts in the parsing algorithm.

Reply:
Yea I was also confused by that, let's descope it out of this CEP for now.